### PR TITLE
ipq40xx minor clock and CPU fixes

### PR DESCRIPTION
**Pull request by dissent1**
Main fixes are:
1. Sleep clock change. Currently it's set to 32768 in upstream kernel, but QSDK defines it as 32000. I believe sleep clock was just copied from ipq806x or similar target. If someone with datasheet could clarify the correct clock - good, if not then I guess QSDK knows it better.
2. Secondary cpu cores bringup procedure. Potentially fixes secondary cores reset issue. https://patchwork.kernel.org/patch/10467835/
Other fixes are minor.

### CODE REVIEW

**Comment by robimarko** [comment]
Does this also affect 4.19?
What is the issue with secondary cores, we already have patch to get them booted.
Part of it was upstreamed in 4.19.

**Comment by dissent1** [comment]
Doh, forgot about 4.19. Yes, it should be also applied to k4.19. I'll add necessary patches.
There are 2 problems with cpu
1. kpss v2 is not designed for ipq40xx afaik. Even QSDK guys instead of switching to kpss v2 made its own routine. I know that weird stuff happens around QSDK, but on this matter I trust it more.
2. There was an issue with kexec when it's needed to restart secondary cores properly.

**Comment by robimarko** [comment]
Ok, so this should also be submitted upstream.
Hm, well its QSDK guys that upstreamed it using kpss v2.

**Comment by dissent1** [comment]
I believe that it's @blogic who submitted the patch.
https://patchwork.kernel.org/patch/10543691/
Don't know why in kernel repo author is different. Anyway for me it's not obvious that M. McClintock is the QSDK guy.

**Comment by robimarko** [comment]
Sorry, I was thinking about using kpss at all.
Well, this is a classical QCA case in upstream.
Good to see this properly fixed, because currently kpss v2 is used upstream, but L2 cache node was never upstreamed and without it only CPU0 will boot.
So, we have a patch localy for it now.

**Comment by chunkeey** [comment]
> Sleep clock change. Currently it's set to 32768 in upstream kernel, but QSDK defines it as 32000. I believe sleep clock was just copied from ipq806x or similar target. If someone with datasheet could clarify the correct clock - good, if not then I guess QSDK knows it better.

This would be really weird, the 32768 Hz is by far "the common" RTC clock crystal frequency. Simply because it's 2^15 1/s so these are the mass-produced commodities. Note: The 32768 Hz crystals are often shortened to just "32Khz" xo, so I am familiar with the confusion around it. 
I too think the datasheet sould be the decider here.

**Comment by robimarko** [comment]
I got access to the summary datasheet, no mention of sleep clock.

**Comment by chunkeey** [comment]
> I got access to the summary datasheet, no mention of sleep clock.

I guess the BOM of a (few?) devices could work as well. Since this is an marked as an "external clock", so there's supposed to be a crystal for it on the board.

File: target/linux/ipq40xx/patches-4.14/090-ipq40xx-fix-high-resolution-timer.patch [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ arch/arm/boot/dts/qcom-ipq4019.dtsi | 1 +
+ 1 file changed, 1 insertion(+)
+
+--- a/arch/arm/boot/dts/qcom-ipq4019.dtsi
++++ b/arch/arm/boot/dts/qcom-ipq4019.dtsi
+@@ -167,6 +167,7 @@
+ 			     <1 4 0xf08>,
+ 			     <1 1 0xf08>;
+ 		clock-frequency = <48000000>;
++		always-on;
```
Comment (chunkeey) [inline target/linux/ipq40xx/patches-4.14/090-ipq40xx-fix-high-resolution-timer.patch]: without (sysrq-q):

```
[   37.079417] Timer List Version: v0.8
[   37.086807] HRTIMER_MAX_CLOCK_BASES: 4
[   37.089470] now at 37076096318 nsecs
[   37.092997]
[   37.096706] cpu: 0
[   37.098151]  clock 0:
[   37.099978]   .base:       cfdb4440
[   37.102319]   .index:      0
[   37.105645]   .resolution: 10000000 nsecs
[   37.108742]   .get_time:
[   37.108753] ktime_get
```

vs with:

```
[12912.405691] Timer List Version: v0.8
[12912.412521] HRTIMER_MAX_CLOCK_BASES: 4
[12912.415763] now at 12912484526672 nsecs
[12912.420041] 
[12912.422989] cpu: 0
[12912.424688]  clock 0:
[12912.426514]   .base:       c7cb4440
[12912.428857]   .index:      0
[12912.432190]   .resolution: 1 nsecs
[12912.435280]   .get_time:   
[12912.435289] ktime_get
[12912.438490] 
```

:+1: Have you sent this upstream?

File: target/linux/ipq40xx/patches-4.14/089-ipq40xx-fix-sleep-clock.patch [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ 1 file changed, 1 insertion(+), 1 deletion(-)
+
+--- a/arch/arm/boot/dts/qcom-ipq4019.dtsi
++++ b/arch/arm/boot/dts/qcom-ipq4019.dtsi
+@@ -141,9 +141,9 @@
+ 	};
+ 
+ 	clocks {
+-		sleep_clk: sleep_clk {
++		sleep_clk: gcc_sleep_clk_src {
```
Comment (chunkeey) [inline target/linux/ipq40xx/patches-4.14/089-ipq40xx-fix-sleep-clock.patch]: before:

```
cat /sys/kernel/debug/clk/clk_orphan_summary
   clock                         enable_cnt  prepare_cnt        rate   accuracy   phase 
----------------------------------------------------------------------------------------
 gcc_wcss5g_rtc_clk                       1            1           0          0 0
 gcc_wcss2g_rtc_clk                       1            1           0          0 0
 gcc_usb3_sleep_clk                       1            1           0          0 0
 gcc_usb2_sleep_clk                       1            1           0          0 0
```

after:

```
cat /sys/kernel/debug/clk/clk_orphan_summary
   clock                         enable_cnt  prepare_cnt        rate   accuracy   phase
----------------------------------------------------------------------------------------
```

:+1:

File: target/linux/ipq40xx/patches-4.14/090-ipq40xx-fix-high-resolution-timer.patch (hunk 1 above)
Comment (dissent1) [reply target/linux/ipq40xx/patches-4.14/090-ipq40xx-fix-high-resolution-timer.patch]: Nope, but I guess it should be upstreamed along with opp voltage and sleep clock when the confusion is resolved :)

File: target/linux/ipq40xx/patches-4.14/089-ipq40xx-fix-sleep-clock.patch (hunk 2 above)
Comment (dissent1) [reply target/linux/ipq40xx/patches-4.14/089-ipq40xx-fix-sleep-clock.patch]: I noticed the wrong name when I tried to find any clue in the driver for the correct sleep clock rate

**Comment by dissent1** [comment]
Patches for k4.19 are in

**Comment by chunkeey** [comment]
@dissent1 I didn't get the chance to test the "add support for secondary cores bringup" this weekend. I know that there was a big performance difference between kpssv1 and kpssv2 when running the openssl benchmark, so I have to postpone this for the next weekend (unless someone else with the hardware does it).

Do you think you could squash the two patches of "add support for secondary cores bringup" together and send them all together with the patches for the frequency and clock to the ML? You can add me to the CC' and once they are up (as in "just" posted, I'm not going to wait for that ML to react) I'll queue everything.

(Note: In some of the patches the "Author" in the commits is dissent1! So watch out before sending them off)

**Comment by dissent1** [comment]
@chunkeey I've squashed CPU patches and fixed the author (thanks for noting!).
Will send the PR to the mailing list shortly.

**Comment by chunkeey** [comment]
Hm, what should happen to this PR now? Should I just close it since newer patches are on the OpenWrt ML?

**Comment by dissent1** [comment]
Yeah :)

**Pull request closed without merging**